File: frontend/src/lib/api/files.ts
import { AnalysisApiClient } from './analysis';
import type { FileAIAnalysis, FileItem, PaginatedResponse } from './types';

export class FilesApiClient extends AnalysisApiClient {
  async downloadFiles(groupId: number | string, maxFiles?: number, sortBy: string = 'download_count',
                     downloadInterval: number = 1.0, longSleepInterval: number = 60.0,
                     filesPerBatch: number = 10, downloadIntervalMin?: number,
                     downloadIntervalMax?: number, longSleepIntervalMin?: number,
                     longSleepIntervalMax?: number) {
    const requestBody: any = {
      max_files: maxFiles,
      sort_by: sortBy,
      download_interval: downloadInterval,
      long_sleep_interval: longSleepInterval,
      files_per_batch: filesPerBatch
    };

    // 如果提供了随机间隔范围参数，则添加到请求中
    if (downloadIntervalMin !== undefined) {
      requestBody.download_interval_min = downloadIntervalMin;
      requestBody.download_interval_max = downloadIntervalMax;
      requestBody.long_sleep_interval_min = longSleepIntervalMin;
      requestBody.long_sleep_interval_max = longSleepIntervalMax;
    }

    return this.request(`/api/files/download/${groupId}`, {
      method: 'POST',
      body: JSON.stringify(requestBody),
    });
  }

  async downloadFilesByTimeRange(groupId: number | string, params: {
    startTime?: string;
    endTime?: string;
    lastDays?: number;
    downloadInterval?: number;
    longSleepInterval?: number;
    filesPerBatch?: number;
    downloadIntervalMin?: number;
    downloadIntervalMax?: number;
    longSleepIntervalMin?: number;
    longSleepIntervalMax?: number;
  }) {
    const requestBody: any = {
      sort_by: 'create_time',
      start_time: params.startTime,
      end_time: params.endTime,
      last_days: params.lastDays,
      download_interval: params.downloadInterval ?? 1.0,
      long_sleep_interval: params.longSleepInterval ?? 60.0,
      files_per_batch: params.filesPerBatch ?? 10,
    };

    if (params.downloadIntervalMin !== undefined) {
      requestBody.download_interval_min = params.downloadIntervalMin;
      requestBody.download_interval_max = params.downloadIntervalMax;
      requestBody.long_sleep_interval_min = params.longSleepIntervalMin;
      requestBody.long_sleep_interval_max = params.longSleepIntervalMax;
    }

    return this.request(`/api/files/download/${groupId}`, {
      method: 'POST',
      body: JSON.stringify(requestBody),
    });
  }

  async collectFiles(groupId: number | string) {
    return this.request(`/api/files/collect/${groupId}`, {
      method: 'POST',
    });
  }

  async collectFilesByTimeRange(groupId: number | string, params: {
    startTime?: string;
    endTime?: string;
    lastDays?: number;
  }) {
    return this.request(`/api/files/collect/${groupId}`, {
      method: 'POST',
      body: JSON.stringify({
        start_time: params.startTime,
        end_time: params.endTime,
        last_days: params.lastDays,
      }),
    });
  }

  async clearFileDatabase(groupId: number | string) {
    return this.request(`/api/files/clear/${groupId}`, {
      method: 'POST',
    });
  }

  async getFileStats(groupId: number | string) {
    return this.request(`/api/files/stats/${groupId}`);
  }

  async syncFilesFromTopics(groupId: number | string): Promise<{
    success: boolean;
    group_id: string;
    stats: {
      scanned: number;
      new_files: number;
      relations: number;
      topic_files: number;
    };
  }> {
    return this.request(`/api/files/sync-from-topics/${groupId}`, {
      method: 'POST',
    });
  }

  async downloadSingleFile(groupId: string, fileId: number, fileName?: string, fileSize?: number) {
    const params = new URLSearchParams();
    if (fileName) params.append('file_name', fileName);
    if (fileSize !== undefined) params.append('file_size', fileSize.toString());

    const url = `/api/files/download-single/${groupId}/${fileId}${params.toString() ? '?' + params.toString() : ''}`;
    return this.request(url, {
      method: 'POST',
    });
  }

  async getFileStatus(groupId: string, fileId: number) {
    return this.request(`/api/files/status/${groupId}/${fileId}`);
  }

  async checkLocalFileStatus(groupId: string, fileName: string, fileSize: number) {
    const params = new URLSearchParams({
      file_name: fileName,
      file_size: fileSize.toString(),
    });
    return this.request(`/api/files/check-local/${groupId}?${params}`);
  }

  async getFiles(groupId: number, page: number = 1, perPage: number = 20, status?: string, search?: string): Promise<PaginatedResponse<FileItem>> {
    const params = new URLSearchParams({
      page: page.toString(),
      per_page: perPage.toString(),
    });

    if (status) {
      params.append('status', status);
    }

    if (search) {
      params.append('search', search);
    }

    const response = await this.request<{files: FileItem[], pagination: any}>(`/api/files/${groupId}?${params}`);
    return {
      data: response.files,
      pagination: response.pagination,
    };
  }

  async getFileAIAnalysis(groupId: number | string, fileId: number): Promise<{ analysis: FileAIAnalysis | null }> {
    return this.request(`/api/files/analysis/${groupId}/${fileId}`);
  }

  async analyzeFile(groupId: number | string, fileId: number, force: boolean = false): Promise<{ analysis: FileAIAnalysis }> {
    return this.request(`/api/files/analysis/${groupId}/${fileId}`, {
      method: 'POST',
      body: JSON.stringify({ force }),
    });
  }
}

File: backend/schemas/files.py
from __future__ import annotations

from typing import Optional

from pydantic import BaseModel, Field


class FileDownloadRequest(BaseModel):
    max_files: Optional[int] = Field(default=None, description="最大下载文件数")
    sort_by: str = Field(default="download_count", description="排序方式: download_count 或 time")
    start_time: Optional[str] = Field(default=None, description="下载时间范围开始时间 YYYY-MM-DD 或 ISO 时间")
    end_time: Optional[str] = Field(default=None, description="下载时间范围结束时间 YYYY-MM-DD 或 ISO 时间")
    last_days: Optional[int] = Field(default=None, ge=1, le=3650, description="下载最近多少天的文件")
    download_interval: float = Field(default=1.0, ge=0.1, le=300.0, description="单次下载间隔（秒）")
    long_sleep_interval: float = Field(default=60.0, ge=10.0, le=3600.0, description="长休眠间隔（秒）")
    files_per_batch: int = Field(default=10, ge=1, le=100, description="下载多少文件后触发长休眠")
    download_interval_min: Optional[float] = Field(default=None, ge=1.0, le=300.0, description="随机下载间隔最小值（秒）")
    download_interval_max: Optional[float] = Field(default=None, ge=1.0, le=300.0, description="随机下载间隔最大值（秒）")
    long_sleep_interval_min: Optional[float] = Field(default=None, ge=10.0, le=3600.0, description="随机长休眠间隔最小值（秒）")
    long_sleep_interval_max: Optional[float] = Field(default=None, ge=10.0, le=3600.0, description="随机长休眠间隔最大值（秒）")


class FileCollectRequest(BaseModel):
    start_time: Optional[str] = Field(default=None, description="收集时间范围开始日期 YYYY-MM-DD")
    end_time: Optional[str] = Field(default=None, description="收集时间范围结束日期 YYYY-MM-DD")
    last_days: Optional[int] = Field(default=None, ge=1, le=3650, description="收集最近多少天的文件")


class FileAIAnalysisRequest(BaseModel):
    force: bool = Field(default=False, description="是否强制重新分析")


class FileIdListRequest(BaseModel):
    file_ids: list[int] = Field(..., min_length=1, max_length=200, description="文件 ID 列表")


class FileAIAnalysisBatchRequest(FileIdListRequest):
    force: bool = Field(default=False, description="是否强制重新分析")


class FileFilteredDownloadRequest(BaseModel):
    status: Optional[str] = Field(default=None, description="文件下载状态筛选")
    search: Optional[str] = Field(default=None, description="文件搜索关键词")
    max_files: Optional[int] = Field(default=None, ge=1, le=5000, description="最多下载多少个匹配文件")
